lib/event_tracer/datadog_logger.rb

```diff
@@ -22,50 +23,33 @@ class DatadogLogger < BasicDecorator
     DEFAULT_METRIC_TYPE = :count
     DEFAULT_COUNTER = 1
 
-    attr_reader :allowed_tags
-
     def initialize(decoratee, allowed_tags: [], default_tags: {})
       super(decoratee)
       @allowed_tags = allowed_tags.freeze
       @default_tags = default_tags.freeze
     end
 
-    LOG_TYPES.each do |log_type|
-      define_method log_type do |**args|
-        metrics = args[:metrics]
-
-        return fail_result('Invalid Datadog config') unless valid_args?(metrics)
-        return success_result if metrics.empty?
+    private
 
-        tags = build_tags(args)
+    def handle_metrics(metrics, args)
+      tags = build_tags(args)
 
-        case metrics
-        when Array
-          metrics.each do |metric|
-            datadog.public_send(DEFAULT_METRIC_TYPE, metric, DEFAULT_COUNTER, tags: tags)
-          end
-        when Hash
-          metrics.each do |metric_name, metric_payload|
-            payload = metric_payload.transform_keys(&:to_sym)
-            metric_type = SUPPORTED_METRIC_TYPES[payload.fetch(:type).to_sym]
-            datadog.public_send(metric_type, metric_name, payload.fetch(:value), tags: tags) if metric_type
-          end
+      case metrics
+      when Array
+        metrics.each do |metric|
+          datadog.public_send(DEFAULT_METRIC_TYPE, metric, DEFAULT_COUNTER, tags: tags)
+        end
+      when Hash
+        metrics.each do |metric_name, metric_payload|
+          payload = metric_payload.transform_keys(&:to_sym)
+          metric_type = SUPPORTED_METRIC_TYPES[payload.fetch(:type).to_sym]
+          datadog.public_send(metric_type, metric_name, payload.fetch(:value), tags: tags) if metric_type
```

Comment:
Could we also move this `case..when..` block to parent class? I think each child class can define only this part:
```
datadog.public_send(metric_type, metric_name, payload.fetch(:value), tags: tags) if metric_type
```
and the metric types mapper.